technology-considerations/graphql-prototypes.md

```diff
@@ -0,0 +1,178 @@
+
+### Local development
+
+It was fairly straightforward to set up an [AppSync
+prototype][tdr-appsync-prototype] based on the [Serverless
+Framework][serverless]. This connected to DynamoDB and just used Velocity to map
+the data.
+
+As mentioned above, we wanted to check that we could write business logic in
+Scala. [Serverless supports Lambdas written in Java][serverless-java] (which
+implies it also supports other JVM languages like Scala), but we weren't able to
+set up a working dev environment that used both Java _and_ AppSync. It may well
+be possible - we just decided it wasn't worth spending more time getting it to
+work.
+
+Writing pipeline resolvers also looks like it involves a lot of boilerplate
+code, since every Lambda resolver requires at least four Velocity templates
+(before and after the pipeline, and before and after each Lambda function).
+We suspect that it would be difficult to maintain a mix of Velocity and pipeline
+resolvers, since some code will inevitably be duplicated.
+
+We looked into [AWS Amplify CLI][amplify] as an alternative, but it doesn't
+appear to support Java Lambdas. It also looks quite rigid: it's designed to help
+people develop a very specific stack (AppSync with React and mobile clients). We
+want something more stack-agnostic so that we can integrate it with other TDR
+services.
+
+Developing an AppSync application without one of these framework looks
+impractical because you would have to either build some way to run the templates
+and the Lambda code together, or give up on local development and only run code
+in an AWS environment.
+
+One final reason disadvantage of AppSync is that it requires you to write Apache
+Velocity templates. One of the developers on the team has used Velocity in
+a previous project and found it very difficult to read, maintain and test.
+
+[tdr-appsync-prototype]: https://github.com/nationalarchives/prototype-appsync
+[serverless]: https://serverless.com/
+[serverless-java]: https://serverless.com/blog/how-to-create-a-rest-api-in-java-using-dynamodb-and-serverless/
+[amplify]: https://github.com/aws-amplify/amplify-cli
+
+### Deployment
+
+The Serverless Framework and Amplify both depend on [CloudFormation], which
+would be tricky to fit in with our [Terraform configuration][tdr-terraform].
+
+If we built our own deployment pipeline, it would have to deploy the Velocity
+templates using the AWS API as well as deploying the Lambda code by copying
+build artifacts.
+
+[CloudFormation]: https://aws.amazon.com/cloudformation/
+[tdr-terraform]: https://github.com/nationalarchives/tdr-prototype-terraform
+
+### Hosting
+
+Using AppSync would make the application extremely coupled to AWS
+Infrastructure, because so much of the code would be in Velocity templates and
+the rest would be implemented as AWS Lambdas.
+
+### Database connection
+
+TODO: Ian or Sam to add details of why it was hard to connect to the DB
```

Comment:
Simple resolvers require an AppSync data source
Aurora (a database that we are considering) is not a supported Data Source, lambda resolvers are required.
Aurora Serverless can be used as a data source when set up to use the Data API. Data API is not available in the London region.
To use Aurora Serverless without the Data API lambdas within a private VPC must be used and these can have a slow cold start.

Reply:
Thanks. :+1: 

I've updated this section - can you check that it makes sense?

Reply:
yes